test: format-truncation warning in dbwrapper_tests

Compiling master at e1bf5470f919cf212703466411968916db8ae61f on Ubuntu 22.10 (with depends):

```cpp
  CXX      test/test_bitcoin-dbwrapper_tests.o
test/dbwrapper_tests.cpp: In member function ‘void dbwrapper_tests::iterator_string_ordering::test_method()’:
test/dbwrapper_tests.cpp:365:41: error: ‘%d’ directive output may be truncated writing between 1 and 11 bytes into a region of size 10 [-Werror=format-truncation=]
  365 |             snprintf(buf, sizeof(buf), "%d", x);
      |                                         ^~
test/dbwrapper_tests.cpp:365:40: note: directive argument in the range [-2147483648, 9]
  365 |             snprintf(buf, sizeof(buf), "%d", x);
      |                                        ^~~~
In file included from /usr/include/stdio.h:906,
                 from /usr/include/c++/12/cstdio:42,
                 from /usr/include/c++/12/ext/string_conversions.h:43,
                 from /usr/include/c++/12/bits/basic_string.h:3960,
                 from /usr/include/c++/12/string:53,
                 from ./clientversion.h:30,
                 from ./dbwrapper.h:8,
                 from test/dbwrapper_tests.cpp:5:
In function ‘int snprintf(char*, size_t, const char*, ...)’,
    inlined from ‘void dbwrapper_tests::iterator_string_ordering::test_method()’ at test/dbwrapper_tests.cpp:365:21:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 2 and 12 bytes into a destination of size 10
   54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   55 |                                    __glibc_objsize (__s), __fmt,
      |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   56 |                                    __va_arg_pack ());
      |                                    ~~~~~~~~~~~~~~~~~
```

(I ran with `--enable-werror`, otherwise it's merely a warning)

cc @achow101

Managed to recreate by making `depends` with `DEBUG=1`, which adds `-D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC -D_LIBCPP_DEBUG=1` to the `CPPFLAGS` when configuring `bitcoin`, then running with your `configure` options. 

This is *really* bizarre. I made a simple test file with the same behavior and tried to compile it on the Ubuntu instance.

```
ubuntu@ip-172-30-1-49:~/bitcoin$ g++ t.cpp -o t.bin -fsanitize=undefined 
ubuntu@ip-172-30-1-49:~/bitcoin$ g++ t.cpp -o t.bin -fsanitize=undefined -O1 
t.cpp: In function ‘int main()’:
t.cpp:11:41: warning: ‘%d’ directive output may be truncated writing between 1 and 11 bytes into a region of size 9 [-Wformat-truncation=]
   11 |             snprintf(buf, sizeof(buf), "%d", x);
      |                                         ^~
t.cpp:11:40: note: directive argument in the range [-2147483648, 9]
   11 |             snprintf(buf, sizeof(buf), "%d", x);
      |                                        ^~~~
In file included from /usr/include/stdio.h:906,
                 from t.cpp:1:
In function ‘int snprintf(char*, size_t, const char*, ...)’,
    inlined from ‘int main()’ at t.cpp:11:21:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 2 and 12 bytes into a destination of size 9
   54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   55 |                                    __glibc_objsize (__s), __fmt,
      |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   56 |                                    __va_arg_pack ());
      |                                    ~~~~~~~~~~~~~~~~~

```

Compiles fine without `-O1`, but gives the warning with it.

Anyway, I can confirm that changing it to an `unsigned int` fixes the issue.

--- a/src/test/dbwrapper_tests.cpp
+++ b/src/test/dbwrapper_tests.cpp
@@ -5,6 +5,7 @@
 #include <dbwrapper.h>
 #include <test/util/setup_common.h>
 #include <uint256.h>
+#include <util/string.h>
 
 #include <memory>
 
@@ -324,12 +325,6 @@ struct StringContentsSerializer {
     StringContentsSerializer() = default;
     explicit StringContentsSerializer(const std::string& inp) : str(inp) {}
 
-    StringContentsSerializer& operator+=(const std::string& s) {
-        str += s;
-        return *this;
-    }
-    StringContentsSerializer& operator+=(const StringContentsSerializer& s) { return *this += s.str; }
-
     template<typename Stream>
     void Serialize(Stream& s) const
     {
@@ -343,44 +338,34 @@ struct StringContentsSerializer {
     {
         str.clear();
         uint8_t c{0};
-        while (true) {
-            try {
-                s >> c;
-                str.push_back(c);
-            } catch (const std::ios_base::failure&) {
-                break;
-            }
+        while (!s.eof()) {
+            s >> c;
+            str.push_back(c);
         }
     }
 };
 
 BOOST_AUTO_TEST_CASE(iterator_string_ordering)
 {
-    char buf[10];
-
     fs::path ph = m_args.GetDataDirBase() / "iterator_string_ordering";
     CDBWrapper dbw(ph, (1 << 20), true, false, false);
-    for (int x=0x00; x<10; ++x) {
-        for (int y = 0; y < 10; y++) {
-            snprintf(buf, sizeof(buf), "%d", x);
-            StringContentsSerializer key(buf);
-            for (int z = 0; z < y; z++)
+    for (int x = 0; x < 10; ++x) {
+        for (int y = 0; y < 10; ++y) {
+            std::string key{ToString(x)};
+            for (int z = 0; z < y; ++z)
                 key += key;
             uint32_t value = x*x;
-            BOOST_CHECK(dbw.Write(key, value));
+            BOOST_CHECK(dbw.Write(StringContentsSerializer{key}, value));
         }
     }
 
     std::unique_ptr<CDBIterator> it(const_cast<CDBWrapper&>(dbw).NewIterator());
     for (const int seek_start : {0, 5}) {
-        snprintf(buf, sizeof(buf), "%d", seek_start);
-        StringContentsSerializer seek_key(buf);
-        it->Seek(seek_key);
-        for (unsigned int x=seek_start; x<10; ++x) {
-            for (int y = 0; y < 10; y++) {
-                snprintf(buf, sizeof(buf), "%d", x);
-                std::string exp_key(buf);
-                for (int z = 0; z < y; z++)
+        it->Seek(StringContentsSerializer{ToString(seek_start)});
+        for (unsigned int x = seek_start; x < 10; ++x) {
+            for (int y = 0; y < 10; ++y) {
+                std::string exp_key{ToString(x)};
+                for (int z = 0; z < y; ++z)
                     exp_key += exp_key;
                 StringContentsSerializer key;
                 uint32_t value;

--- a/test/lint/lint-locale-dependence.py
+++ b/test/lint/lint-locale-dependence.py
@@ -45,7 +45,6 @@ from subprocess import check_output, CalledProcessError
 
 KNOWN_VIOLATIONS = [
     "src/dbwrapper.cpp:.*vsnprintf",
-    "src/test/dbwrapper_tests.cpp:.*snprintf",
     "src/test/fuzz/locale.cpp:.*setlocale",
     "src/test/fuzz/string.cpp:.*strtol",
     "src/test/fuzz/string.cpp:.*strtoul",

